cirq-google/cirq_google/engine/abstract_local_processor.py

```diff
@@ -0,0 +1,410 @@
+        """Returns True if the slot is available for reservation."""
+        for t in self._schedule:
+            if t.slot_type == qenums.QuantumTimeSlot.TimeSlotType.UNALLOCATED:
+                continue
+            if t.end_time.seconds and t.end_time.seconds <= time_slot.start_time.seconds:
+                continue
+            if t.start_time.seconds and t.start_time.seconds >= time_slot.end_time.seconds:
+                continue
+            return False
+        return True
+
+    def create_reservation(
+        self,
+        start_time: datetime.datetime,
+        end_time: datetime.datetime,
+        whitelisted_users: Optional[List[str]] = None,
+    ) -> qtypes.QuantumReservation:
+        """Creates a reservation on this processor.
+
+        Args:
+            start_time: the starting date/time of the reservation.
+            end_time: the ending date/time of the reservation.
+            whitelisted_users: a list of emails that are allowed
+              to send programs during this reservation (in addition to users
+              with permission "quantum.reservations.use" on the project).
+
+        Raises:
+            ValueError: if start_time is after end_time.
+        """
+        if end_time < start_time:
+            raise ValueError('End time of reservation must be after the start time')
+        reservation_id = self._create_id()
+        new_reservation = qtypes.QuantumReservation(
+            name=reservation_id,
+            start_time=Timestamp(seconds=int(start_time.timestamp())),
+            end_time=Timestamp(seconds=int(end_time.timestamp())),
+            whitelisted_users=whitelisted_users,
+        )
+        time_slot = self._reservation_to_time_slot(new_reservation)
+        if not self._is_available(time_slot):
+            raise ValueError('Time slot is not available for reservations')
+
+        self._reservations[reservation_id] = new_reservation
+        self._insert_reservation_into(time_slot)
+        return new_reservation
+
+    def remove_reservation(self, reservation_id: str):
+        """Removes a reservation on this processor."""
+        if reservation_id in self._reservations:
+            del self._reservations[reservation_id]
+
+    def get_reservation(self, reservation_id: str):
+        """Retrieve a reservation given its id."""
+        if reservation_id in self._reservations:
+            return self._reservations[reservation_id]
+        else:
+            return None
+
+    def update_reservation(
+        self,
+        reservation_id: str,
+        start_time: datetime.datetime = None,
+        end_time: datetime.datetime = None,
+        whitelisted_users: List[str] = None,
+    ):
+        """Updates a reservation with new information.
+
+        Updates a reservation with a new start date, end date, or
+        list of additional users.  For each field, it the argument is left as
+        None, it will not be updated.
```

Comment:
Missing `Args` and `Raises` annotations.

Reply:
`end_time: New starting time` -> `end_time: New ending time`

Reply:
Fixed.